### PR TITLE
storage: manage sink connections when rendering

### LINKED ISSUE #17210: sentry: panic: Reference to absent collection: u13762
Sentry Issue: [DATABASE-BACKEND-2H7](https://sentry.io/organizations/materializeinc/issues/3883334435/?referrer=github_integration)

Filed by @antiguru. This doesn't seem to happen frequently, but in compute it was an indication of a race condition somewhere. Might be a good idea to keep this on the radar.

```
panic: Reference to absent collection: u13762
  File "panicking.rs", line 575, in rust_begin_unwind
  File "client.rs", line 521, in mz_storage_client::client::PartitionedStorageState<T>::absorb_response
  File "client.rs", line 148, in <mz_service::client::Partitioned<mz_service::local::LocalClient<mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageResponse, std::thread::Thread>, mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageRes...
  File "select.rs", line 517, in mz_service::grpc::GrpcServer<T>::forward_bidi_stream::
  File "poll_fn.rs", line 58, in <tokio::future::poll_fn::PollFn<<mz_service::grpc::GrpcServer<mz_storage::server::serve::{closure#4}>>::forward_bidi_stream<mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageResponse, mz_storage_client::client::ProtoStorageCom...
...
(86 additional frame(s) were not displayed)
```

ci-regexp: Reference to absent collection

### LINKED ISSUE #20019: [Inc 67] sinks: bootstrapping catalog does not wait for sink creation
### What version of Materialize are you using?

main

### What is the issue?

Part of #17210

Bootstrapping the coordinator does not wait for sinks to be created, which leads to inconsistent state. While this is a problem generally, it specifically leads to a pathological panic if the storage controller attempts to reconnect to a running worker (known as reconciliation).

1. During bootstrapping, we [defer the work of building the connection](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/adapter/src/coord.rs#L1107-L1128), which has is an asynchronous process that we ostensibly do not want to hold up starting Materialize.
2. The coordinator notifies the storage controller that [initialization completed](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/adapter/src/coord.rs#LL1213C14-L1213C14), which [kicks off the process](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage-client/src/controller.rs#LL1217C20-L1217C20) of [reconciliation](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L972-L977) in the the individual worker's storage state.
	Important to note that from the perspective of the storage controller, we have only generated a source export token and the lower-level clients (e.g. the rehydrating storage client for the cluster where the sink runs) have absolutely no record of the sink's pending creation.
3. During reconciliation, the storage state assumes [all running exports are stale](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L986) and removes them from being marked as stale if it [sees a `CreateSinks` command with the ID](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L1072). In this case, reconciliation occurs while the sink's connection is still getting created, so we don't know that it shouldn't be reconciled away––and then reconcile it away.

This means we are almost certain to drop sinks running on existing workers.

## Misc. thoughts

I was very surprised to see that the sink code path does not rely on purification like sources do. My first inclination is that this code should pattern itself after sources and push down the problem of maintaining connection health into storage.

### LINKED ISSUE #20020: storage: reconciliation on running workers leads to panics
### What version of Materialize are you using?

main

### What is the issue?

Part of #17210

1. During reconciliation, we might decide to reconcile away (i.e. [drop](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L1108-L1115) some running dataflows that are no longer in the controller we're agreeing to work with.
2. We process these synthesized drop commands [just as we would any other command](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L1122-L1130), i.e. there is no special flag that these IDs are being dropped by reconciliation.
3. Whenever we receive an [`AllowCompaction` to the empty frontier](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L1268-L1287) (a drop), we [track it as a dropped ID](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L856).
4. We periodically [report our dropped IDs](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage/src/storage_state.rs#L521C22-L525) to the PartitionedState to which we're connected, at which time we've forgotten the provenance of the drop.
5. In the case of reconciliation, we explicitly and only drop IDs which the client driving us doesn't have. If we do see a command, the partitioned client [tracks the uppers](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage-client/src/client.rs#L541-L546).
6. When the worker sends back a response that it has dropped the IDs it reconciled away, [the PartitionedState has no record of the collection](https://github.com/MaterializeInc/materialize/blob/52c1cac3cd37640e0a73cb7468b39111c50b9a6a/src/storage-client/src/client.rs#L649-L656), leading to the panic.

## Why don't we hit this more often

It is a long-known issue that we do not properly emit `DroppedId` responses (#8185), so triggering this requires reconciliation on an active worker with an apparent catalog change in the move between epochs. I don't think that scenario is well tested, but the repeated occurrence of #17210 suggests it's not so rare.

## Misc.
- Fix seems easy but testing seems hard.
- This bug alerted us to the existence of #20019

### PR DESCRIPTION
**Pull request by sploiselle**
Establish Kafka broker connections + topics when rendering sinks, rather than during sequencing.

@MaterializeInc/testing I could use some help developing tests for this. Namely, I think we could use a test that:
1. Sets up a sink, produces data, etc.
2. Reboots MZ but leaves the worker running the sink.
3. Verifies that the cluster that was running the sink did not panic. This is currently the case on `main` because we fail to reconcile the running cluster's sinks with the catalog.

Closes #17210
Closes #20019
Closes #20020

### Motivation

 This PR adds a known-desirable feature. #20919

### Tips for reviewer

The methods that need the most review are:

- `build_kafka` in `src/storage-client/src/sink.rs`, which now sets up the collateral topics in the Kafka broker during rendering (whereas previously these were done during sequencing).
- `encode_stream` in `src/storage/src/sink/kafka.rs`, which now holds `build_kafka`.

The following commits need review:

- `storage/sink: error if progress data present, but data topic missing`
- `storage: allow reconciling runnings sinks w/ differing as ofs`, which looks a little goofy but will make more sense in the context of `ALTER CONNECTION`, which I have a sketch PR currentlu open for #21239

I was also unsure of the most handsome way to handle the evolution of the `KafkaSinkConnection` proto. It's really an evolution of the `KafkaSinkConnectionBuilder`, but is different enough that the number of reserved fields was obnoxious, so wanted to start afresh. Appending `V2` feels not quite right, but wasn't sure if there's a better approach.

The remainder of the code is either uninteresting code movement, deleting unused code paths, or somewhat straightforward work that doesn't require close scrutiny (e.g. purification).

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by sploiselle** [comment]
I will be out until Oct 16, but if anyone wants to take this over please do.

- Source and sink error reporting is most likely to be broken due to the checks for consistency data in the progress topic. I've iterated on it a few times and it's most consistent than it was but might not be fool-proof.
- Other misc. errors are most likely to be caused by string formatting or "merge skew" for tests that don't wait for sink topic creation after a successful `CREATE SINK`.
- This needs at least one test to ensure that MZ can reboot with a sink running on a cluster and successfully reconcile the cluster with a new envd (i.e. it doesn't panic). This test should likely introduce some delay between the Kafka broker and MZ to "prove" #https://github.com/MaterializeInc/materialize/issues/20019 is fixed.

**Comment by philip-stoev** [comment]
Since this PR is supposed to fix the `Reference to absent collection` , and I have a test that runs into that particular panic using unified clusters, I ran the test against this PR. Unfortunately the workloads now panic with the `New export with same ID` panic. https://buildkite.com/materialize/nightlies/builds/4540#_

**Comment by philip-stoev** [comment]
Also, several tests in this PR are failing as follows:

```


2023-09-29 18:11:15 UTC | > CREATE SINK rt_binding_progress_test_sink FROM rt_binding_progress_test_source
-- | --
  | 2023-09-29 18:11:15 UTC | INTO KAFKA CONNECTION kafka_conn (TOPIC 'rt-binding-progress-test-output-1538756022')
  | 2023-09-29 18:11:15 UTC | FORMAT AVRO
  | 2023-09-29 18:11:15 UTC | USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn ENVELOPE DEBEZIUM
  | 2023-09-29 18:11:16 UTC | rows match; continuing at ts 1696011076.1280503
  | 2023-09-29 18:11:16 UTC | Verifying results in Kafka topic rt-binding-progress-test-output-1538756022
  | 2023-09-29 18:11:16 UTC | waiting for Kafka topic creation...
  | 2023-09-29 18:13:16 UTC | Received error from Kafka stream consumer: deadline has elapsed
  | 2023-09-29 18:13:16 UTC | kafka-exactly-once-sink.td:320:1: error: fetching schema: subject not found
  | 2023-09-29 18:13:16 UTC | \|
```

This indicates that neither the sink Kafka topic nor the schema registry entry were created 2 min after the `CREATE SINK` statement succeeded.

File: src/sql/src/pure/error.rs [hunk 1]
```diff
@@ -203,3 +203,49 @@ impl LoadGeneratorSourcePurificationError {
...
+
+/// Logical errors detectable during purification for a KAFKA SOURCE.
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum KafkaSinkPurificationError {
+    #[error("{0} is not a KAFKA CONNECTION")]
+    NotKafkaConnection(FullItemName),
+    #[error("admin client errored")]
+    AdminClientError(String),
+    #[error("zero brokers discovered in metadata request")]
+    ZeroBrokers,
```
Comment (guswynn) [inline src/sql/src/pure/error.rs:215]: If you are very careful with thiserror, you can use `Arc<anyhow::Error>` as a useful variant: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9aedb95c48bb501214f7d6e8fba15b7d

**Review by guswynn — COMMENTED** [review]
This is a great idea, and frankly im amazed we did it any other way!

**Comment by sploiselle** [comment]
@philip-stoev Can you link me to a branch that contains the offending test? I'm glad to triage that as part of this work if that turns out to be sane to do.

**Comment by sploiselle** [comment]
@philip-stoev Please see [storage: allow reconciling runnings sinks w/ differing as ofs](https://github.com/MaterializeInc/materialize/pull/21852/commits/2caf2e226dc6bd59a14f87c71117b3d47f74cbb5), which I think resolves the reconciliation issues in a way that are at least satisfying enough to merge. The fix will look less out of place once we support `ALTER CONNECTION`.

@MaterializeInc/storage Can I get a review on this? Glad to walk someone through the code if there are any questions.

**Comment by philip-stoev** [comment]
Let me rerun my tests on top of your commit.

**Comment by philip-stoev** [comment]
@sploiselle Can you please check our this failure:

https://buildkite.com/materialize/nightlies/builds/4785#018b4ca0-8ac0-4ab2-a66e-c85f967b4847

```


> CREATE SINK sink_9
--
  | 2023-10-20 10:30:49 UTC | IN CLUSTER default
  | 2023-10-20 10:30:49 UTC | FROM view_1
  | 2023-10-20 10:30:49 UTC | INTO KAFKA CONNECTION sink_9_kafka_conn (TOPIC 'sink-sink_9')
  | 2023-10-20 10:30:49 UTC | FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION sink_9_csr_conn
  | 2023-10-20 10:30:49 UTC | ENVELOPE DEBEZIUM;
  | 2023-10-20 10:30:49 UTC | rows match; continuing at ts 1697797849.5496292
  | 2023-10-20 10:30:49 UTC | Verifying Kafka topic sink-sink_9
  | 2023-10-20 10:30:49 UTC | waiting for Kafka topic creation...
  | 2023-10-20 10:30:50 UTC | > CREATE SOURCE sink_9_source
  | 2023-10-20 10:30:50 UTC | IN CLUSTER storage
  | 2023-10-20 10:30:50 UTC | FROM KAFKA CONNECTION sink_9_kafka_conn (TOPIC 'sink-sink_9')
  | 2023-10-20 10:30:50 UTC | FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION sink_9_csr_conn
  | 2023-10-20 10:30:50 UTC | ENVELOPE NONE
  | 2023-10-20 10:30:50 UTC |  
  | 2023-10-20 10:30:50 UTC | 16:1: error: executing query failed: db error: ERROR: No value schema found: ERROR: No value schema found

```

So basically `verify-topic` in testdrive confirms that the topic exists but apparently the schema has not been uploaded to the schema registry by the time we attempt to read the topic using `CREATE SOURCE`?  Is that expected? Either tests need to be fortified further to wait for the schema to come up or the schema needs to be uploaded before the topic is created?

**Comment by sploiselle** [comment]
@philip-stoev That is "expected" from the perspective of how the code is currently written. However, it seems like that is going to be a real pain ergonomically, so let me see what we can do to ensure the topic exists _and_ has a schema.

**Comment by sploiselle** [comment]
@philip-stoev Fixed the issue you identified by letting us optionally specify that we want to wait for the value or key schemas, as well. Let me know how this fares.

File: doc/developer/testdrive.md [hunk 2]
```diff
@@ -787,6 +787,13 @@ once one record matches, the following must all match.  There are permitted to b
 topic after the matching is complete.  Note that if the topic is not required to have `partial-search`
 elements in it but there will be an attempt to read up to this number with a blocking read.
 
+#### `kafka-verify-topic [sink=... | topic=...] [await-value-schema=false] [await-key-schema=false]`
```
Comment (moulimukherjee) [inline doc/developer/testdrive.md:790]: OOC can this be used for the progress topic as well?

File: test/source-sink-errors/mzcompose.py [hunk 3]
```diff
@@ -230,7 +231,103 @@ def assert_recovery(self, c: Composition) -> None:
...
+                  ENVELOPE NONE
+                # WITH ( REMOTE 'clusterd:2100' ) https://github.com/MaterializeInc/materialize/issues/16582
 
+                > CREATE SINK sink1 FROM source1
+                  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink-topic-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE DEBEZIUM
+                # WITH ( REMOTE 'clusterd:2100' ) https://github.com/MaterializeInc/materialize/issues/16582
+
+                $ kafka-verify-data format=avro sink=materialize.public.sink1 sort-messages=true
```
Comment (moulimukherjee) [inline test/source-sink-errors/mzcompose.py:305]: Is it possible to add a test to check the contents of the progress topic?

File: test/source-sink-errors/mzcompose.py [hunk 4]
```diff
@@ -203,15 +213,6 @@ def assert_error(self, c: Composition, error: str) -> None:
                   FROM mz_internal.mz_source_statuses
                   WHERE name = 'source1'
                 stalled true
-
-                # Sinks generally halt after receiving an error, which means that they may alternate
-                # between `stalled` and `starting`. Instead of relying on the current status, we
-                # check that there is a stalled status with the expected error.
-                > SELECT bool_or(error ~* '{error}'), bool_or(details->'namespaced'->>'kafka' ~* '{error}')
```
Comment (moulimukherjee) [inline test/source-sink-errors/mzcompose.py:210]: OOC why is this removed? It's validating the error message contents and that the errors are namespaced. Since it's stalled, there would be some corresponding error to check for here?

File: src/storage-types/src/sinks.rs [hunk 5]
```diff
@@ -44,6 +44,63 @@ pub struct StorageSinkDesc<S: StorageSinkDescFillState, T = mz_repr::Timestamp>
...
+    StorageSinkDesc<S, T>
+{
+    /// Determines if `self` is compatible with another `StorageSinkDesc`, in
+    /// such a way that it is possible to turn `self` into `other` through a
+    /// valid series of transformations.
+    ///
+    /// Currently, the only "valid transformation" is the passage of time such
+    /// that the sink's as ofs may differ. However, this will change once we
+    /// support `ALTER CONNECTION` or `ALTER SINK`.
+    pub fn alter_compatible(
```
Comment (moulimukherjee) [inline src/storage-types/src/sinks.rs:57]: If I am reading this right, this method is just validating compatibility, not altering anything yet (which it might do later)?

File: doc/developer/testdrive.md (hunk 2 above)
Comment (sploiselle) [reply doc/developer/testdrive.md:790]: If you specify the topic name rather than specifying the sink, yes

File: test/source-sink-errors/mzcompose.py (hunk 4 above)
Comment (sploiselle) [reply test/source-sink-errors/mzcompose.py:210]: This just moved elsewhere in the same file

File: src/storage-types/src/sinks.rs (hunk 5 above)
Comment (sploiselle) [reply src/storage-types/src/sinks.rs:57]: Yes––this is meant to be used as a preemptive check or an assert that the transformations only occur in expected ways.

File: test/source-sink-errors/mzcompose.py (hunk 3 above)
Comment (sploiselle) [reply test/source-sink-errors/mzcompose.py:305]: This might be tricky to write a test for because we store, I believe, MZ times in the progress topic. If that is the case, I'm not sure how we'd express a wildcard match in the validation here. We can maybe accomplish it with a variable in the testdrive file, but I'm not sure where we could pull it from in such a way that it wouldn't flake.

Will look at this tomorrow but lmk if you have ideas.

**Review by philip-stoev — APPROVED** [review]
@sploiselle the "cluster unification" Zippytest is not happy, apologies for the delay.

Please merge this PR as you see fit and I will get the test in CI afterwards. Thank you!

File: test/source-sink-errors/mzcompose.py (hunk 3 above)
Comment (sploiselle) [reply test/source-sink-errors/mzcompose.py:305]: This is more annoying than I thought because I don't think we publish the progress topic schema to the schema registry service (we always know what it is), but verifying the data requires a published schema. This test feels like it would be nice, but not necessary in the context of this PR--the changes to the progress topic are well tested.

How would you feel about creating a ticket and deferring this work?

Comment (moulimukherjee) [reply test/source-sink-errors/mzcompose.py:305]: sgtm! Thanks for looking into this nevertheless. I think we can tinker with a `kafka-verify-progress-data` or something similar specific to the progress topic.

File: test/source-sink-errors/mzcompose.py (hunk 4 above)
Comment (moulimukherjee) [reply test/source-sink-errors/mzcompose.py:210]: Saw it's in the sink specific disruption 🤦‍♀️

**Review by moulimukherjee — APPROVED** [review]
lgtm!

File: test/source-sink-errors/mzcompose.py (hunk 3 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:305]: FWIW, the progress topic doesn't have a schema in that sense... it's stored as JSON. I haven't tried this, but I would expect `kafka-verify-data format=json` to at least have access to the data. (Though the data in that topic can be unpredictable for other reasons!)
